aten/src/ATen/native/Activation.cpp

```diff
@@ -676,124 +679,60 @@ TORCH_IMPL_FUNC(threshold_backward_out)(const Tensor& grad, const Tensor& self,
+    "prelu: Expected `weight` to be a scalar or 1D tensor, but got: ndim = ", weight_.dim());
+  // Adjust weight to broadcast over self
+  auto weight = weight_.dim() == 0 ? weight_.unsqueeze(0) : weight_;
+  if (self.dim() > 0) {
+    for (C10_UNUSED const auto _ : c10::irange(self.dim() - 2)) {
+      weight = weight.unsqueeze(-1);
     }
-    return t.as_strided(sizes, strides);
-  };
-  Tensor w;
-  if (self.scalar_type() == ScalarType::BFloat16) {
-    auto w_bf16 = at::empty(weight_.sizes(), weight_.options().dtype(ScalarType::BFloat16));
-    w_bf16.copy_(weight_);
-    w = as_nd(w_bf16);
   } else {
-    w = as_nd(weight_);
+    weight = weight.squeeze(0);
   }
-
-  auto iter = TensorIteratorConfig()
-    .add_output(result)
-    .add_input(self)
-    .add_input(w)
-    .build();
-  prelu_cpu_stub(iter.device_type(), iter);
-  return result;
+  return at::_prelu_kernel(self, weight);
 }
 
-std::tuple<Tensor, Tensor> prelu_backward_cpu(const Tensor& grad_out_, const Tensor& self, const Tensor& weight_) {
-  int64_t weight_num = weight_.numel();
-
-  Tensor input_grad = at::empty_like(self, self.suggest_memory_format());
-  Tensor weight_grad = at::empty_like(weight_, at::MemoryFormat::Contiguous);
-  Tensor weight_grad_collector = at::empty_like(self, at::MemoryFormat::Contiguous);
-
-  if (weight_num != 1) {
-    int64_t input_ndim = self.dim();
-    TORCH_CHECK(input_ndim > 0, "Not allow zero-dim input tensor.");
-
-    int64_t channel_size = 1; // channel_size default to 1
-    if (input_ndim > 1) {
-      channel_size = self.size(1); // channel is the 2nd dim of input
-    }
-    TORCH_CHECK(channel_size == weight_num,
-      "Mismatch of parameter numbers and input channel size. Found parameter numbers = ", weight_num,
-      " and channel size = ", channel_size, ".");
-  }
 
-  const int64_t ndim = self.dim();
-  // Helper to convert 1d tensor or scalar tensor to an nd tensor that broadcasts with input
-  // All elements go into the channel dimension
-  DimVector sizes(ndim, 1), strides(ndim, 0);
-  auto as_nd = [&](const Tensor& t) {
-    TORCH_INTERNAL_ASSERT(t.defined() && (t.dim() == 1 || t.dim() == 0));
-    if (ndim >= 2) {
-      sizes[1] = t.dim() == 1 ? t.sizes()[0] : 1;
-      strides[1] = t.dim() == 1 ? t.strides()[0] : 0;
-      return t.as_strided(sizes, strides);
-    }
-    return t.as_strided(sizes, strides);
-  };
-  Tensor w;
-  if (self.scalar_type() == ScalarType::BFloat16) {
-    auto w_bf16 = at::empty(weight_.sizes(), weight_.options().dtype(ScalarType::BFloat16));
-    w_bf16.copy_(weight_);
-    w = weight_.defined() ? as_nd(w_bf16) :
-        at::detail::scalar_tensor_static(1, self.scalar_type(), kCPU);
-  } else {
-    w = weight_.defined() ? as_nd(weight_) :
-        at::detail::scalar_tensor_static(1, self.scalar_type(), kCPU);
-  }
+Tensor _prelu_kernel(const Tensor& self, const Tensor& weight) {
+  auto options = self.options().dtype(promoteTypes(self.scalar_type(), weight.scalar_type()));
```

Comment:
I'm not sure we want type promoting behavior for `prelu`, and we don't have it currently.